share/man/man4/ip.4
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)ip.4	8.2 (Berkeley) 11/30/93
.\" $FreeBSD: src/share/man/man4/ip.4,v 1.13.2.9 2002/05/02 02:40:26 silby Exp $
.\" $DragonFly: src/share/man/man4/ip.4,v 1.7 2007/12/15 10:23:58 swildner Exp $
.\"
.Dd March 3, 2001
.Dt IP 4
.Os
.Sh NAME
.Nm ip
.Nd Internet Protocol
.Sh SYNOPSIS
.In sys/types.h
.In sys/socket.h
.In netinet/in.h
.Ft int
.Fn socket AF_INET SOCK_RAW proto
.Sh DESCRIPTION
.Tn IP
is the transport layer protocol used
by the Internet protocol family.
Options may be set at the
.Tn IP
level
when using higher-level protocols that are based on
.Tn IP
(such as
.Tn TCP
and
.Tn UDP ) .
It may also be accessed
through a
.Dq raw socket
when developing new protocols, or
special-purpose applications.
.Pp
There are several
.Tn IP-level
.Xr setsockopt 2
and
.Xr getsockopt 2
options.
.Dv IP_OPTIONS
may be used to provide
.Tn IP
options to be transmitted in the
.Tn IP
header of each outgoing packet
or to examine the header options on incoming packets.
.Tn IP
options may be used with any socket type in the Internet family.
The format of
.Tn IP
options to be sent is that specified by the
.Tn IP
protocol specification (RFC 791), with one exception:
the list of addresses for Source Route options must include the first-hop
gateway at the beginning of the list of gateways.
The first-hop gateway address will be extracted from the option list
and the size adjusted accordingly before use.
To disable previously specified options,
use a zero-length buffer:
.Bd -literal
setsockopt(s, IPPROTO_IP, IP_OPTIONS, NULL, 0);
.Ed
.Pp
.Dv IP_TOS
and
.Dv IP_TTL
may be used to set the type-of-service and time-to-live
fields in the
.Tn IP
header for
.Dv SOCK_STREAM , SOCK_DGRAM ,
and certain types of
.Dv SOCK_RAW
sockets.
For example,
.Bd -literal
int tos = IPTOS_LOWDELAY;       /* see <netinet/ip.h> */
setsockopt(s, IPPROTO_IP, IP_TOS, &tos, sizeof(tos));

int ttl = 60;                   /* max = 255 */
setsockopt(s, IPPROTO_IP, IP_TTL, &ttl, sizeof(ttl));
.Ed
.Pp
.Dv IP_MINTTL
may be used to set the minimum acceptable TTL a packet must have when
received on a socket.
All packets with a lower TTL are silently dropped.
Works on already connected/connecting and listening sockets for RAW/UDP/TCP.
It allows to implement security mechanisms described in RFC 3682 (GTSM).
.Pp
If the
.Dv IP_RECVDSTADDR
option is enabled on a
.Dv SOCK_DGRAM
socket,
the
.Xr recvmsg 2
call will return the destination
.Tn IP
address for a
.Tn UDP
datagram.
The msg_control field in the msghdr structure points to a buffer
that contains a cmsghdr structure followed by the
.Tn IP
address.
The cmsghdr fields have the following values:
.Bd -literal
cmsg_len = sizeof(struct in_addr)
cmsg_level = IPPROTO_IP
cmsg_type = IP_RECVDSTADDR
.Ed
.Pp
If the
.Dv IP_RECVTTL
option is enabled on a
.Dv SOCK_DGRAM
socket, the
.Xr recvmsg 2
call will return the
.Tn IP
.Tn TTL
(time to live) field for a
.Tn UDP
datagram.
The msg_control field in the msghdr structure points to a buffer
that contains a cmsghdr structure followed by the
.Tn TTL .
The cmsghdr fields have the following values:
.Bd -literal
cmsg_len = sizeof(u_char)
cmsg_level = IPPROTO_IP
cmsg_type = IP_RECVTTL
.Ed
.Pp
.Dv IP_PORTRANGE
may be used to set the port range used for selecting a local port number
on a socket with an unspecified (zero) port number.
It has the following
possible values:
.Bl -tag -width IP_PORTRANGE_DEFAULT
.It Dv IP_PORTRANGE_DEFAULT
use the default range of values, normally
.Dv IPPORT_RESERVED
through
.Dv IPPORT_USERRESERVED .
This is adjustable through the sysctl setting:
.Va net.inet.ip.portrange.first
and
.Va net.inet.ip.portrange.last .
.It Dv IP_PORTRANGE_HIGH
use a high range of values, normally
.Dv IPPORT_HIFIRSTAUTO
and
.Dv IPPORT_HILASTAUTO .
This is adjustable through the sysctl setting:
.Va net.inet.ip.portrange.hifirst
and
.Va net.inet.ip.portrange.hilast .
.It Dv IP_PORTRANGE_LOW
use a low range of ports, which are normally restricted to
privileged processes on
.Ux
systems.  The range is normally from
.Dv IPPORT_RESERVED - 1
down to
.Li IPPORT_RESERVEDSTART
in descending order.
This is adjustable through the sysctl setting:
.Va net.inet.ip.portrange.lowfirst
and
.Va net.inet.ip.portrange.lowlast .
.El
.Ss "Multicast Options"
.Tn IP
multicasting is supported only on
.Dv AF_INET
sockets of type
.Dv SOCK_DGRAM
and
.Dv SOCK_RAW ,
and only on networks where the interface
driver supports multicasting.
.Pp
The
.Dv IP_MULTICAST_TTL
option changes the time-to-live (TTL)
for outgoing multicast datagrams
in order to control the scope of the multicasts:
.Bd -literal
u_char ttl;	/* range: 0 to 255, default = 1 */
setsockopt(s, IPPROTO_IP, IP_MULTICAST_TTL, &ttl, sizeof(ttl));
.Ed
.Pp
Datagrams with a TTL of 1 are not forwarded beyond the local network.
Multicast datagrams with a TTL of 0 will not be transmitted on any network,
but may be delivered locally if the sending host belongs to the destination
group and if multicast loopback has not been disabled on the sending socket
(see below).  Multicast datagrams with TTL greater than 1 may be forwarded
to other networks if a multicast router is attached to the local network.
.Pp
For hosts with multiple interfaces, each multicast transmission is
sent from the primary network interface.
The
.Dv IP_MULTICAST_IF
option overrides the default for
subsequent transmissions from a given socket:
.Bd -literal
struct in_addr addr;
setsockopt(s, IPPROTO_IP, IP_MULTICAST_IF, &addr, sizeof(addr));
.Ed
.Pp
where "addr" is the local
.Tn IP
address of the desired interface or
.Dv INADDR_ANY
to specify the default interface.
An interface's local IP address and multicast capability can
be obtained via the
.Dv SIOCGIFCONF
and
.Dv SIOCGIFFLAGS
ioctls.
Normal applications should not need to use this option.
.Pp
If a multicast datagram is sent to a group to which the sending host itself
belongs (on the outgoing interface), a copy of the datagram is, by default,
looped back by the IP layer for local delivery.
The
.Dv IP_MULTICAST_LOOP
option gives the sender explicit control
over whether or not subsequent datagrams are looped back:
.Bd -literal
u_char loop;	/* 0 = disable, 1 = enable (default) */
setsockopt(s, IPPROTO_IP, IP_MULTICAST_LOOP, &loop, sizeof(loop));
.Ed
.Pp
This option
improves performance for applications that may have no more than one
instance on a single host (such as a router daemon), by eliminating
the overhead of receiving their own transmissions.  It should generally not
be used by applications for which there may be more than one instance on a
single host (such as a conferencing program) or for which the sender does
not belong to the destination group (such as a time querying program).
.Pp
A multicast datagram sent with an initial TTL greater than 1 may be delivered
to the sending host on a different interface from that on which it was sent,
if the host belongs to the destination group on that other interface.  The
loopback control option has no effect on such delivery.
.Pp
A host must become a member of a multicast group before it can receive
datagrams sent to the group.  To join a multicast group, use the
.Dv IP_ADD_MEMBERSHIP
option:
.Bd -literal
struct ip_mreq mreq;
setsockopt(s, IPPROTO_IP, IP_ADD_MEMBERSHIP, &mreq, sizeof(mreq));
.Ed
.Pp
where
.Fa mreq
is the following structure:
.Bd -literal
struct ip_mreq {
    struct in_addr imr_multiaddr; /* IP multicast address of group */
    struct in_addr imr_interface; /* local IP address of interface */
}
.Ed
.Pp
.Fa imr_interface
should
be
.Dv INADDR_ANY
to choose the default multicast interface,
or the
.Tn IP
address of a particular multicast-capable interface if
the host is multihomed.
Membership is associated with a single interface;
programs running on multihomed hosts may need to
join the same group on more than one interface.
Up to
.Dv IP_MAX_MEMBERSHIPS
(currently 20) memberships may be added on a
single socket.
.Pp
To drop a membership, use:
.Bd -literal
struct ip_mreq mreq;
setsockopt(s, IPPROTO_IP, IP_DROP_MEMBERSHIP, &mreq, sizeof(mreq));
.Ed
.Pp
where
.Fa mreq
contains the same values as used to add the membership.
Memberships are dropped when the socket is closed or the process exits.
.\"-----------------------
.Ss "Raw IP Sockets"
Raw
.Tn IP
sockets are connectionless,
and are normally used with the
.Xr sendto 2
and
.Xr recvfrom 2
calls, though the
.Xr connect 2
call may also be used to fix the destination for future
packets (in which case the
.Xr read 2
or
.Xr recv 2
and
.Xr write 2
or
.Xr send 2
system calls may be used).
.Pp
If
.Fa proto
is 0, the default protocol
.Dv IPPROTO_RAW
is used for outgoing
packets, and only incoming packets destined for that protocol
are received.
If
.Fa proto
is non-zero, that protocol number will be used on outgoing packets
and to filter incoming packets.
.Pp
Outgoing packets automatically have an
.Tn IP
header prepended to
them (based on the destination address and the protocol
number the socket is created with),
unless the
.Dv IP_HDRINCL
option has been set.
Incoming packets are received with
.Tn IP
header and options intact.
.Pp
.Dv IP_HDRINCL
indicates the complete IP header is included with the data
and may be used only with the
.Dv SOCK_RAW
type.
.Bd -literal
#include <netinet/in_systm.h>
#include <netinet/ip.h>

int hincl = 1;                  /* 1 = on, 0 = off */
setsockopt(s, IPPROTO_IP, IP_HDRINCL, &hincl, sizeof(hincl));
.Ed
.Pp
Unlike previous
.Bx
releases, the program must set all
the fields of the IP header, including the following:
.Bd -literal
ip->ip_v = IPVERSION;
ip->ip_hl = hlen >> 2;
ip->ip_id = 0;  /* 0 means kernel set appropriate value */
ip->ip_off = offset;
.Ed
.Pp
If the header source address is set to
.Dv INADDR_ANY ,
the kernel will choose an appropriate address.
.Sh ERRORS
A socket operation may fail with one of the following errors returned:
.Bl -tag -width Er
.It Bq Er EISCONN
when trying to establish a connection on a socket which
already has one, or when trying to send a datagram with the destination
address specified and the socket is already connected;
.It Bq Er ENOTCONN
when trying to send a datagram, but
no destination address is specified, and the socket hasn't been
connected;
.It Bq Er ENOBUFS
when the system runs out of memory for
an internal data structure;
.It Bq Er EADDRNOTAVAIL
when an attempt is made to create a
socket with a network address for which no network interface
exists.
.It Bq Er EACCES
when an attempt is made to create
a raw IP socket by a non-privileged process.
.El
.Pp
The following errors specific to
.Tn IP
may occur when setting or getting
.Tn IP
options:
.Bl -tag -width EADDRNOTAVAILxx
.It Bq Er EINVAL
An unknown socket option name was given.
.It Bq Er EINVAL
The IP option field was improperly formed;
an option field was shorter than the minimum value
or longer than the option buffer provided.
.El
.Sh SEE ALSO
.Xr getsockopt 2 ,
.Xr recv 2 ,
.Xr send 2 ,
.Xr icmp 4 ,
.Xr inet 4 ,
.Xr intro 4
.Sh HISTORY
The
.Nm
protocol appeared in
.Bx 4.2 .


usr.bin/systat/ip.c
/*-
 * Copyright (c) 1980, 1992, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * @(#)mbufs.c	8.1 (Berkeley) 6/6/93
 * $FreeBSD: src/usr.bin/systat/ip.c,v 1.3.2.1 2001/04/25 12:42:18 ru Exp $
 * $DragonFly: src/usr.bin/systat/ip.c,v 1.5 2004/05/03 15:18:25 hmp Exp $
 */

#include <sys/param.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/sysctl.h>

#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <netinet/ip.h>
#include <netinet/ip_var.h>
#include <netinet/udp.h>
#include <netinet/udp_var.h>

#include <stdlib.h>
#include <string.h>
#include <paths.h>
#include "systat.h"
#include "extern.h"
#include "mode.h"

struct stat {
	struct ip_stats i;
	struct udpstat u;
};

static struct stat curstat, initstat, oldstat;

/*-
--0         1         2         3         4         5         6         7
--0123456789012345678901234567890123456789012345678901234567890123456789012345
01          IP Input                           IP Output
02999999999 total packets received   999999999 total packets sent
03999999999 - with bad checksums     999999999 - generated locally
04999999999 - too short for header   999999999 - output drops
05999999999 - too short for data     999999999 output fragments generated
06999999999 - with invalid hlen      999999999 - fragmentation failed
07999999999 - with invalid length    999999999 destinations unreachable
08999999999 - with invalid version   999999999 packets output via raw IP
09999999999 - jumbograms
10999999999 total fragments received           UDP Statistics
11999999999 - fragments dropped      999999999 total input packets
12999999999 - fragments timed out    999999999 - too short for header
13999999999 - packets reassembled ok 999999999 - invalid checksum
14999999999 packets forwarded        999999999 - no checksum
15999999999 - unreachable dests      999999999 - invalid length
16999999999 - redirects generated    999999999 - no socket for dest port
17999999999 option errors            999999999 - no socket for broadcast
18999999999 unwanted multicasts      999999999 - socket buffer full
19999999999 delivered to upper layer 999999999 total output packets
--0123456789012345678901234567890123456789012345678901234567890123456789012345
--0         1         2         3         4         5         6         7
*/

WINDOW *
openip(void)
{
	return (subwin(stdscr, LINES-4-1, 0, 4, 0));
}

void
closeip(WINDOW *w)
{
	if (w == NULL)
		return;
	wclear(w);
	wrefresh(w);
	delwin(w);
}

void
labelip(void)
{
	wmove(wnd, 0, 0); wclrtoeol(wnd);
#define L(row, str) mvwprintw(wnd, row, 10, str)
#define R(row, str) mvwprintw(wnd, row, 45, str);
	L(1, "IP Input");			R(1, "IP Output");
	L(2, "total packets received");		R(2, "total packets sent");
	L(3, "- with bad checksums");		R(3, "- generated locally");
	L(4, "- too short for header");		R(4, "- output drops");
	L(5, "- too short for data");	R(5, "output fragments generated");
	L(6, "- with invalid hlen");	R(6, "- fragmentation failed");
	L(7, "- with invalid length");	R(7, "destinations unreachable");
	L(8, "- with invalid version");	R(8, "packets output via raw IP");
	L(9, "- jumbograms");
	L(10, "total fragments received");	R(10, "UDP Statistics");
	L(11, "- fragments dropped");	R(11, "total input packets");
	L(12, "- fragments timed out");	R(12, "- too short for header");
	L(13, "- packets reassembled ok");	R(13, "- invalid checksum");
	L(14, "packets forwarded");	R(14, "- no checksum");
	L(15, "- unreachable dests");	R(15, "- invalid length");
	L(16, "- redirects generated");	R(16, "- no socket for dest port");
	L(17, "option errors");		R(17, "- no socket for broadcast");
	L(18, "unwanted multicasts");	R(18, "- socket buffer full");
	L(19, "delivered to upper layer");	R(19, "total output packets");
#undef L
#undef R
}

static void
domode(struct stat *ret)
{
	const struct stat *sub;
	double divisor = 1.0;

	switch(currentmode) {
	case display_RATE:
		sub = &oldstat;
		divisor = naptime;
		break;
	case display_DELTA:
		sub = &oldstat;
		break;
	case display_SINCE:
		sub = &initstat;
		break;
	default:
		*ret = curstat;
		return;
	}
#define DO(stat) ret->stat = (double)(curstat.stat - sub->stat) / divisor
	DO(i.ips_total);
	DO(i.ips_badsum);
	DO(i.ips_tooshort);
	DO(i.ips_toosmall);
	DO(i.ips_badhlen);
	DO(i.ips_badlen);
	DO(i.ips_fragments);
	DO(i.ips_fragdropped);
	DO(i.ips_fragtimeout);
	DO(i.ips_forward);
	DO(i.ips_cantforward);
	DO(i.ips_redirectsent);
	DO(i.ips_noproto);
	DO(i.ips_delivered);
	DO(i.ips_localout);
	DO(i.ips_odropped);
	DO(i.ips_reassembled);
	DO(i.ips_fragmented);
	DO(i.ips_ofragments);
	DO(i.ips_cantfrag);
	DO(i.ips_badoptions);
	DO(i.ips_noroute);
	DO(i.ips_badvers);
	DO(i.ips_rawout);
	DO(i.ips_toolong);
	DO(i.ips_notmember);
	DO(u.udps_ipackets);
	DO(u.udps_hdrops);
	DO(u.udps_badsum);
	DO(u.udps_nosum);
	DO(u.udps_badlen);
	DO(u.udps_noport);
	DO(u.udps_noportbcast);
	DO(u.udps_fullsock);
	DO(u.udps_opackets);
#undef DO
}
	
void
showip(void)
{
	struct stat stats;
	u_long totalout;

	domode(&stats);
	totalout = stats.i.ips_forward + stats.i.ips_localout;

#define DO(stat, row, col) \
	mvwprintw(wnd, row, col, "%9lu", stats.stat)

	DO(i.ips_total, 2, 0);
	mvwprintw(wnd, 2, 35, "%9lu", totalout);
	DO(i.ips_badsum, 3, 0);
	DO(i.ips_localout, 3, 35);
	DO(i.ips_tooshort, 4, 0);
	DO(i.ips_odropped, 4, 35);
	DO(i.ips_toosmall, 5, 0);
	DO(i.ips_ofragments, 5, 35);
	DO(i.ips_badhlen, 6, 0);
	DO(i.ips_cantfrag, 6, 35);
	DO(i.ips_badlen, 7, 0);
	DO(i.ips_noroute, 7, 35);
	DO(i.ips_badvers, 8, 0);
	DO(i.ips_rawout, 8, 35);
	DO(i.ips_toolong, 9, 0);
	DO(i.ips_fragments, 10, 0);
	DO(i.ips_fragdropped, 11, 0);
	DO(u.udps_ipackets, 11, 35);
	DO(i.ips_fragtimeout, 12, 0);
	DO(u.udps_hdrops, 12, 35);
	DO(i.ips_reassembled, 13, 0);
	DO(u.udps_badsum, 13, 35);
	DO(i.ips_forward, 14, 0);
	DO(u.udps_nosum, 14, 35);
	DO(i.ips_cantforward, 15, 0);
	DO(u.udps_badlen, 15, 35);
	DO(i.ips_redirectsent, 16, 0);
	DO(u.udps_noport, 16, 35);
	DO(i.ips_badoptions, 17, 0);
	DO(u.udps_noportbcast, 17, 35);
	DO(i.ips_notmember, 18, 0);
	DO(u.udps_fullsock, 18, 35);
	DO(i.ips_delivered, 19, 0);
	DO(u.udps_opackets, 19, 35);
#undef DO
}

#define CPU_STATS_FUNC(proto,type)                            \
static void                                                   \
proto ##_stats_agg(type *ary, type *ttl, int cpucnt)          \
{                                                             \
    int i, off, siz;                                          \
    siz = sizeof(type);                                       \
                                                              \
    if (!ary && !ttl)                                         \
        return;                                               \
                                                              \
    bzero(ttl, siz);                                          \
    if (cpucnt == 1) {                                        \
        *ttl = ary[0];                                        \
    } else {                                                  \
        for (i = 0; i < cpucnt; ++i) {                        \
            for (off = 0; off < siz; off += sizeof(u_long)) { \
                *(u_long *)((char *)(*(&ttl)) + off) +=       \
                *(u_long *)((char *)&ary[i] + off);           \
            }                                                 \
        }                                                     \
    }                                                         \
}
CPU_STATS_FUNC(ip, struct ip_stats);

static int
fetch_ipstats(struct ip_stats *st)
{
	struct ip_stats stattmp[SMP_MAXCPU];
	size_t len;
	int name[4], cpucnt;

	name[0] = CTL_NET;
	name[1] = PF_INET;
	name[2] = IPPROTO_IP;
	name[3] = IPCTL_STATS;

	len = sizeof(struct ip_stats) * SMP_MAXCPU;
	if (sysctl(name, 4, stattmp, &len, NULL, 0) < 0) {
		error("sysctl getting ip_stats failed");
		return -1;
	}
	cpucnt = len / sizeof(struct ip_stats);
	ip_stats_agg(stattmp, st, cpucnt);

	return 0;
}

int
initip(void)
{
	size_t len;
	int name[4];

	if (fetch_ipstats(&initstat.i) < 0)
		return 0;

	name[0] = CTL_NET;
	name[1] = PF_INET;
	name[2] = IPPROTO_UDP;
	name[3] = UDPCTL_STATS;

	len = 0;
	if (sysctl(name, 4, 0, &len, 0, 0) < 0) {
		error("sysctl getting udpstat size failed");
		return 0;
	}
	if (len > sizeof curstat.u) {
		error("ip_stats structure has grown--recompile systat!");
		return 0;
	}
	if (sysctl(name, 4, &initstat.u, &len, 0, 0) < 0) {
		error("sysctl getting udpstat failed");
		return 0;
	}
	oldstat = initstat;
	return 1;
}

void
resetip(void)
{
	size_t len;
	int name[4];

	fetch_ipstats(&initstat.i);

	name[0] = CTL_NET;
	name[1] = PF_INET;
	name[2] = IPPROTO_UDP;
	name[3] = UDPCTL_STATS;

	len = sizeof initstat.u;
	if (sysctl(name, 4, &initstat.u, &len, 0, 0) < 0) {
		error("sysctl getting udpstat failed");
	}
	oldstat = initstat;
}

void
fetchip(void)
{
	int name[4];
	size_t len;

	oldstat = curstat;

	if (fetch_ipstats(&curstat.i) < 0)
		return;

	name[0] = CTL_NET;
	name[1] = PF_INET;
	name[2] = IPPROTO_UDP;
	name[3] = UDPCTL_STATS;
	len = sizeof curstat.u;

	if (sysctl(name, 4, &curstat.u, &len, 0, 0) < 0)
		return;
}
